src/macros.jl

```diff
@@ -237,7 +237,126 @@ macro grad(expr)
+`ReverseDiff.TrackedReal` and `ReverseDiff.TrackedArray`
+respectively. For example, we can import `rrule` of `f(::Real,
+::Array)` like below:
+
+
+```
+ReverseDiff.@grad_from_chainrules f(x::TrackedReal, ::TrackedArray)
+ReverseDiff.@grad_from_chainrules f(x::TrackedReal, ::Array)
+ReverseDiff.@grad_from_chainrules f(x::Real, ::TrackedArray)
+```
+
+"""
+macro grad_from_chainrules(fcall)
+    @gensym tp output_value output back closure cls_args cls_kwargs
+    f, fwd_def = funcall_fwd_to_rule(fcall)
+
+    return quote
+        $fwd_def
+        function $ReverseDiff.track(::typeof($f), args...; kwargs...)
+            $tp = $ReverseDiff.tape(args...)
+            $output_value, $back = ChainRules.rrule($f, map(ReverseDiff.value, args)...; kwargs...)
+            $output = $ReverseDiff.track($output_value, $tp)
+            $closure($cls_args...; $cls_kwargs...) = ChainRules.rrule($f, map(ReverseDiff.value, $cls_args)...; $cls_kwargs...)
+            $ReverseDiff.record!(
+                $tp,
+                $ReverseDiff.SpecialInstruction,
+                $f,
+                args,
+                $output,
+                ($back, $closure, kwargs),
+            )
+            return $output
+        end
+
+        if !hasmethod(
+            $ReverseDiff.special_reverse_exec!,
+            Tuple{$ReverseDiff.SpecialInstruction{typeof($f)}},
+        )
+            @noinline function $ReverseDiff.special_reverse_exec!(instruction::$ReverseDiff.SpecialInstruction{typeof($f)})
+                output = instruction.output
+                input = instruction.input
+                back = instruction.cache[1]
+                back_output = back($ReverseDiff.deriv(output))
+                input_derivs = back_output[2:end]
+                @assert input_derivs isa Tuple
+                $ReverseDiff._add_to_deriv!.(input, input_derivs)
+                $ReverseDiff.unseed!(output)
+                return nothing
+            end
+        end
+
+        if !hasmethod(
+            $ReverseDiff.special_forward_exec!,
+            Tuple{$ReverseDiff.SpecialInstruction{typeof($f)}},
+        )
+            @noinline function $ReverseDiff.special_forward_exec!(instruction::$ReverseDiff.SpecialInstruction{typeof($f)})
+                output, input = instruction.output, instruction.input
+                $ReverseDiff.pull_value!.(input)
+                pullback = instruction.cache[2]
+                kwargs = instruction.cache[3]
+                out_value = pullback(input...; kwargs...)[1]
+                $ReverseDiff.value!(output, out_value)
+                return nothing
+            end
+        end
+    end |> esc
+end
+
 _add_to_deriv!(d1, d2) = nothing
+function _add_to_deriv!(d1::Union{TrackedReal, AbstractArray{<:TrackedReal}}, t::InplaceableThunk)
+    _add_to_deriv!(d1, t.val)
+end
+function _add_to_deriv!(d1::Union{TrackedReal, AbstractArray{<:TrackedReal}}, t::Thunk)
+    increment_deriv!(d1, unthunk(t))
+end
```

Comment:
Maybe a more orthogonal design would be better, i.e., maybe we could just use `unthunk` in the calls to `_add_to_deriv!` above instead of adding more dispatches. E.g,
```julia
_add_to_deriv!.(input, unthunk.(input_derivs))
```
instead of  `_add_to_deriv!.(input, input_derivs)`. Alternatively, if something more general is needed, we could define a function such as
```julia
unwrap_chainrules_output(x) = x
unwrap_chainrules_output(x::AbstractThunk) = unthunk(x)
...
```
and use it in the call to `_add_to_deriv!`.

In any case, we should not specialize on `Thunk` or `InplaceableThunk` but on `AbstractThunk`, and not access internal fields but just use `unthunk`.

Reply:
updated.